pkg/ttnpb/application.pb.validate.go

```diff
@@ -0,0 +1,1005 @@
+
+// ErrorName returns error name.
+func (e ApplicationValidationError) ErrorName() string { return "ApplicationValidationError" }
+
+// Error satisfies the builtin error interface
+func (e ApplicationValidationError) Error() string {
+	cause := ""
+	if e.cause != nil {
+		cause = fmt.Sprintf(" | caused by: %v", e.cause)
+	}
+
+	key := ""
+	if e.key {
+		key = "key for "
+	}
+
+	return fmt.Sprintf(
+		"invalid %sApplication.%s: %s%s",
+		key,
+		e.field,
+		e.reason,
+		cause)
+}
+
+var _ error = ApplicationValidationError{}
+
+var _ interface {
+	Field() string
+	Reason() string
+	Key() bool
+	Cause() error
+	ErrorName() string
+} = ApplicationValidationError{}
+
+// ValidateFields checks the field values on Applications with the rules
+// defined in the proto definition for this message. If any rules are
+// violated, an error is returned.
+func (m *Applications) ValidateFields(paths ...string) error {
+	if m == nil {
+		return nil
+	}
+
+	if len(paths) == 0 {
+		paths = ApplicationsFieldPathsNested
+	}
+
+	for name, subs := range _processPaths(append(paths[:0:0], paths...)) {
+		_ = subs
+		switch name {
+		case "applications":
+
+			for idx, item := range m.GetApplications() {
+				_, _ = idx, item
+
+				if v, ok := interface{}(item).(interface{ ValidateFields(...string) error }); ok {
+					if err := v.ValidateFields(subs...); err != nil {
+						return ApplicationsValidationError{
+							field:  fmt.Sprintf("applications[%v]", idx),
+							reason: "embedded message failed validation",
+							cause:  err,
+						}
+					}
+				}
+
+			}
+
+		default:
+			return ApplicationsValidationError{
+				field:  name,
+				reason: "invalid field path",
+			}
+		}
+	}
+	return nil
+}
+
+// ApplicationsValidationError is the validation error returned by
+// Applications.ValidateFields if the designated constraints aren't met.
+type ApplicationsValidationError struct {
+	field  string
+	reason string
+	cause  error
+	key    bool
+}
+
+// Field function returns field value.
+func (e ApplicationsValidationError) Field() string { return e.field }
+
+// Reason function returns reason value.
+func (e ApplicationsValidationError) Reason() string { return e.reason }
+
+// Cause function returns cause value.
+func (e ApplicationsValidationError) Cause() error { return e.cause }
+
+// Key function returns key value.
+func (e ApplicationsValidationError) Key() bool { return e.key }
+
+// ErrorName returns error name.
+func (e ApplicationsValidationError) ErrorName() string { return "ApplicationsValidationError" }
+
+// Error satisfies the builtin error interface
+func (e ApplicationsValidationError) Error() string {
+	cause := ""
+	if e.cause != nil {
+		cause = fmt.Sprintf(" | caused by: %v", e.cause)
+	}
+
+	key := ""
+	if e.key {
+		key = "key for "
+	}
+
+	return fmt.Sprintf(
+		"invalid %sApplications.%s: %s%s",
+		key,
+		e.field,
+		e.reason,
+		cause)
+}
+
+var _ error = ApplicationsValidationError{}
```

Comment:
These kind of assertions belong in test code (same below). For every XXXValidationError we initialize two, each taking up 32 bytes of memory (or so). Do we really need to generate them? Can't we just add a unit test in the generator to make sure that (any) generated XXXValidationError implements these?

Reply:
IDEM, see https://github.com/TheThingsNetwork/lorawan-stack/pull/194#discussion_r261240612